8297247: Add GarbageCollectorMXBean for Remark and Cleanup pause time in G1

linade opened the pull request:
Hi,

This change uses a G1MonitorScope (covers jstat and GC beans) to replace TraceCollectorStats (only covers jstat) in VM_G1PauseConcurrent::doit(), so that now Remark and Cleanup phase time are both collected by a GC bean.
This change adds a third bean "G1 Concurrent GC", after "G1 Young Generation" and "G1 Old generation". This has good continuity, meaning that any current application code who uses "G1 Young Generation" and "G1 Old generation" will see nothing changed, but it will now observe a third bean and can decide what to do with it, e.g., summing all of them up.
Also now the gc beans and jstat fields have 1-to-1 correspondance:
"G1 Young Generation" -- YGC/YGCT
"G1 Old generation" -- FGC/FGCT
"G1 Concurrent GC" -- CGC/CGCT

Alternative:
I've considered using the existing bean "G1 Old generation", but it's currently used for Full GCs, and I find unifying Full GCs with Remark and Cleanup under the same bean is a little awkward. For example, I couldn't find a proper name for the G1MemoryManager, or a proper "gc end message" for the post-gc notification. Also, in this way, if users have previously hooked "G1 Old generation" to some kind of Full GC alarm, then they might suddenly find a lot of false-positives (caused by Remark and Cleanup). So this is undesirable.

Ran tests:
test/hotspot/jtreg:tier1-4
test/jdk:tier1
test/jdk:jdk_jfr_sanity
test/jdk:needs_g1gc
-Failure found and fixed: jdk/jfr/event/gc/collection/TestGCEventMixedWithG1ConcurrentMark.java
-Failure found and fixed: jdk/jfr/event/gc/collection/TestGCEventMixedWithG1FullCollection.java
test/jdk:jdk_management
-Failure found and fixed: com/sun/management/GarbageCollectorMXBean/GarbageCollectionNotificationContentTest.java
-Failure found and fixed: com/sun/management/GarbageCollectorMXBean/GarbageCollectionNotificationTest.java
-Failure found and fixed: java/lang/management/MemoryMXBean/MemoryTest.java#id0
test/hotspot/jtreg:vmTestbase_vm_gc
test/hotspot/jtreg:vmTestbase_nsk_monitoring
test/hotspot/jtreg:vmTestbase_vm_gc_misc
test/jdk:jdk_jmx
test/hotspot/jtreg:hotspot_containers_extended

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297247](https://bugs.openjdk.org/browse/JDK-8297247): Add GarbageCollectorMXBean for Remark and Cleanup pause time in G1


### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11341/head:pull/11341` \
`$ git checkout pull/11341`

Update a local copy of the PR: \
`$ git checkout pull/11341` \
`$ git pull https://git.openjdk.org/jdk pull/11341/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11341`

View PR using the GUI difftool: \
`$ git pr show -t 11341`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11341.diff">https://git.openjdk.org/jdk/pull/11341.diff</a>

</details>

tschatzl commented on src/hotspot/share/gc/g1/g1MonitoringSupport.cpp:
```diff
@@ -344,15 +348,32 @@ MemoryUsage G1MonitoringSupport::old_gen_memory_usage(size_t initial_size, size_
-  _tms(full_gc ? &monitoring_support->_full_gc_memory_manager : &monitoring_support->_incremental_memory_manager,
-       G1CollectedHeap::heap()->gc_cause(), all_memory_pools_affected) {
-}
+G1MonitoringScope::G1MonitoringScope(G1MonitoringSupport* monitoring_support)
+: _monitoring_support(monitoring_support) { }
```
```suggestion
G1MonitoringScope::G1MonitoringScope(G1MonitoringSupport* monitoring_support) :
  _monitoring_support(monitoring_support) { }
```

tschatzl commented on src/hotspot/share/gc/g1/g1MonitoringSupport.hpp:
```diff
@@ -239,11 +239,32 @@ class G1MonitoringSupport : public CHeapObj<mtGC> {
+  ~G1MonitoringScope();
+};
+
+class G1IncMonitoringScope : public G1MonitoringScope {
```
```suggestion
class G1YoungGCMonitoringScope : public G1MonitoringScope {
```
Everything about young collection is prefixed with "YoungGC", let's keep naming consistent.

tschatzl commented on src/hotspot/share/gc/g1/g1MonitoringSupport.hpp:
```diff
@@ -239,11 +239,32 @@ class G1MonitoringSupport : public CHeapObj<mtGC> {
+  ~G1IncMonitoringScope() { }
+};
+
+class G1FullMonitoringScope : public G1MonitoringScope {
```
```suggestion
class G1FullGCMonitoringScope : public G1MonitoringScope {
```
Similiarly, the prefix should be "FullGC" for consistency.

tschatzl commented on src/hotspot/share/gc/g1/g1MonitoringSupport.hpp:
```diff
@@ -239,11 +239,32 @@ class G1MonitoringSupport : public CHeapObj<mtGC> {
+  ~G1FullMonitoringScope() { }
 };
 
+class G1ConcMonitoringScope : public G1MonitoringScope {
```
```suggestion
class G1ConcGCMonitoringScope : public G1MonitoringScope {
```
The member is called `_conc_gc_...` and the name of the bean also "Concurrent GC...".

tschatzl reviewed (changes requested):
Lgtm, apart from some naming consistency issues.

tschatzl commented:
I'll run it through our testing too.

linade commented:
> I'll run the change through our testing too, taking a bit.

Thanks for the review! I appended the list of tests I ran. The beans are used in many other tests. I will cover them over the weekend and get back to you.

linade commented:
Updated with suggested fixes. I missed some test failures earlier by accident. I fixed them and it should be fine now. Appended the failures found and fixed since last review. In particular, there is a destruction order between G1MonitoringScope and TraceMemoryManagerStats that must be respected. Didn't catch that earlier.

tschatzl reviewed (changes requested):
Some additional comments, mainly on the tests.
This change also needs a release note to announce the new GC MXBean.

tschatzl commented:
I also changed the CR description to "Add GarbageCollectorMXBean for Remark and Cleanup pause time in G1" since this is what this change does (and helps with the release note naming). The bot will complain about the PR/CR naming difference, please fix.

linade commented on src/hotspot/share/gc/g1/g1MonitoringSupport.hpp:
```diff
@@ -122,12 +122,16 @@ class MemoryPool;
   G1CollectedHeap* _g1h;
 
   // java.lang.management MemoryManager and MemoryPool support
   GCMemoryManager _incremental_memory_manager;
```
Yes, probably also for other memory_managers? in a separate CR.

linade commented on test/jdk/com/sun/management/GarbageCollectorMXBean/GarbageCollectionNotificationContentTest.java:
```diff
@@ -100,6 +106,14 @@ public static void main(String[] args) throws Exception {
         for(int i = 0; i<10000000; i++) {
             data[i%32] = new int[8];
         }
```
This test also tests if the allocated objects are collected. So there still needs to be a few allocations. Not sure if we can replace it.

linade commented on test/hotspot/jtreg/gc/g1/TestRemarkCleanupMXBean.java:
```diff
@@ -0,0 +1,79 @@
+                System.out.println(bean.getName() + " reports a difference " +
+                                   after + " - " + before + " = " + (after - before));
+            }
+        }
```
done

linade commented on test/jdk/com/sun/management/GarbageCollectorMXBean/GarbageCollectionNotificationTest.java:
```diff
@@ -99,6 +105,14 @@ public static void main(String[] args) throws Exception {
         for(int i = 0; i<100000000; i++) {
             data[i%32] = new int[8];
         }
```
I find this class MixedGCProvoker does what we need. I made a minor modification to it.

linade commented:
> Some additional comments, mainly on the tests. This change also needs a release note to announce the new GC MXBean.

Do you think the latest commit addresses your concerns? I put up a release note here https://bugs.openjdk.org/browse/JDK-8297247
please review : )

tschatzl commented:
> I put up a release note here https://bugs.openjdk.org/browse/JDK-8297247 please review : )

I did not really understand the sentence "As a `MemoryManagerMXBean`, this bean manages the "G1 Old Gen" memory pool." and changed it to "These pauses now also update the "G1 Old Gen" `MemoryManagerMXBean` memory pool."

Other than that I added a few newlines here and there to form paragraphs.

linade commented:
/integrate

linade commented:
Thank you @tschatzl . Are you able to sponsor this change?

tschatzl commented:
Sure, but we'll need a second reviewer. I'll ping others.

albertnetymk commented on src/hotspot/share/gc/g1/g1MonitoringSupport.cpp:
```diff
@@ -356,3 +363,23 @@ G1MonitoringScope::~G1MonitoringScope() {
+G1YoungGCMonitoringScope::G1YoungGCMonitoringScope(G1MonitoringSupport* monitoring_support,
+                                                   bool all_memory_pools_affected) :
+  G1MonitoringScope(monitoring_support,
+                    monitoring_support->_incremental_collection_counters,
+                    &monitoring_support->_incremental_memory_manager,
```
Preexisting: it's slightly odd that one is a pointer and the other not. I wonder if both can be non-pointer.

linade replied:
This can be considered in another CR I think, together with the renaming of these variables, proposed in an earlier comment.

albertnetymk commented on src/hotspot/share/gc/g1/g1MonitoringSupport.hpp:
```diff
@@ -242,8 +241,28 @@ class G1MonitoringScope : public StackObj {
+class G1YoungGCMonitoringScope : public G1MonitoringScope {
+public:
+  G1YoungGCMonitoringScope(G1MonitoringSupport* monitoring_support, bool all_memory_pools_affected);
+  ~G1YoungGCMonitoringScope() { }
```
I think the empty destructor in child classes can be omitted, and the destructor in base class can probably be `protected`.

linade replied:
done

albertnetymk commented on test/hotspot/jtreg/gc/g1/TestRemarkCleanupMXBean.java:
```diff
@@ -0,0 +1,74 @@
+ * @bug 8297247
+ * @summary Test that Remark and Cleanup are correctly reported by
+ *          a GarbageCollectorMXBean
+ * @requires vm.gc.G1 & vm.compMode != "Xcomp"
```
Why `vm.compMode != "Xcomp"`?

linade replied:
We were afraid that '-Xcomp' might bring some disruptive GCs to (an earlier version of) the test. Removed it, considering now that test is simplified.

tschatzl commented:
/sponsor

linade commented:
/integrate

linade commented:
@tschatzl do you mind sponsor again?

tschatzl commented:
/sponsor

linade commented:
Thanks @tschatzl @albertnetymk . I'll create a second bug id for the minor issues brought up during the review.
